mlflow/gateway/providers/togetherai.py

```diff
@@ -0,0 +1,428 @@
+        # ":[DONE]
+
+        return completions_schema.StreamResponsePayload(
+            id=resp.get("id"),
+            created=resp.get("created"),
+            model=config.model.name,
+            choices=[
+                completions_schema.StreamChoice(
+                    index=idx,
+                    # TODO this is questionable since the finish reason comes from togetherai api
+                    finish_reason=choice.get("finish_reason"),
+                    delta=completions_schema.StreamDelta(role=None, content=choice.get("text")),
+                )
+                for idx, choice in enumerate(resp.get("choices", []))
+            ],
+            # usage is not included in OpenAI StreamResponsePayload
+            # usage=completions.CompletionsUsage(
+            #    prompt_tokens=usage.get("prompt_tokens") if usage else None,
+            #    input_tokens=usage.get("completion_tokens") if usage else None,
+            #    total_tokens=usage.get("total_tokens") if usage else None,
+            # )
+        )
+
+    @classmethod
+    def completions_to_model(cls, payload, config):
+        key_mapping = {
+            # Parameter that share common names or functionalities in OpenAI and TogetherAI
+            # TogetherAI parameter range for repetition penalty
+            # repetition_penalty (-3.402823669209385e+38, 3.402823669209385e+38)
+            # OpenAI parameter range for frequency penalty and presence penalty
+            # frequency_penalty (-2.0, 2.0)
+            # presence_penalty (-2.0, 2.0)
+            "frequency_penalty": "repetition_penalty",
+            # OpenAI parameter range for top_logprobs
+            # top_logprobs (0, 20)
+            # TogetherAI parameter range for logprobs
+            # logprobs (-2147483648,2147483648)
+            "top_logprobs": "logprobs",
+        }
+        # if user tries to use the OpenAI api format
+        # logprobs parameter in OpenAI is a boolean flag
+        # if logprobs is present in the payload top_logprobs should be specified
+        # condition one: check if logprobs is in payload
+        # if it is check if its set to true, since its boolean
+        logprobs_in_payload_condition = (
+            "logprobs" in payload and isinstance(payload["logprobs"], bool) and payload["logprobs"]
+        )
+        # if the above are true then top_logprobs should contain a value
+        # if it's not, it's wrong and logprobs should have been specified to false
+        # taken by the API reference (https://platform.openai.com/docs/api-reference/chat/create)
+        openai_logprobs_condition = logprobs_in_payload_condition and (
```

Comment:
If my understanding is correct, TogheterAI doesn't have equivalent to `OpenAI's logprobs` and TogetherAI's `logprobs` parameter is OpenAI's `top_logprobs` right? In that case, would it be better if we can just only accept `logprobs` as `integer` value following TogetherAI's definition? Then if the user passes boolean value, we raise exception and say "please specify integer values....".

The general rule of thumb is that we should **not** change the query parameters aside from the [standard parameters](https://mlflow.org/docs/latest/llms/deployments/index.html#completions). Taking [HuggingFace provider](https://github.com/mlflow/mlflow/blob/b6556c8a25aad1c94f370732d8752c73f898cdaf/mlflow/gateway/providers/huggingface.py#L39-L41) as an example, we map standard parameters like `max_tokens`, `temperature`, `max_tokens`, and `temperature` to their definition. There are [much more parameters](https://huggingface.co/docs/transformers/en/main_classes/text_generation#transformers.GenerationConfig) can be specified to HuggingFace interface (including `repetition_penalty`) but we accept them as they are rather than mapping it to OpenAI format.

The main reason is that we don't know exact implementation of the providers and the information we can get from public documentations is limited. Hence parameter mappings often includes some guess, which has risk of confusion even when documented properly. Also considering the number of possible parameters and providers, it is not realistic for us to maintain all parameters and change mapping logic upon underlying provider's update🙂

Reply:
> In that case, would it be better if we can just only accept logprobs as integer value following TogetherAI's definition? Then if the user passes boolean value, we raise exception and say "please specify integer values....".

Yeap, and that would mean if they pass ``top_logprobs`` , we allow for the end API to throw an error? 


> The main reason is that we don't know exact implementation of the providers and the information we can get from public documentations is limited. Hence parameter mappings often includes some guess, which has risk of confusion even when documented properly. Also considering the number of possible parameters and providers, it is not realistic for us to maintain all parameters and change mapping logic upon underlying provider's update🙂

Okay that makes sense to me!

Reply:
```python
@classmethod
def completions_to_model(cls, payload, config):

    key_mapping = {
        # TogetherAI uses logprobs
        # OpenAI uses top_logprobs
        "top_logprobs": "logprobs"
    }

    # in openAI API the logprobs parameter 
    # is a boolean flag. 
    # Insert this here to prevent the user from mixing up the APIs
    logprobs_in_payload_condition = (
        "logprobs" in payload and not isinstance(payload["logprobs"], int)
    )

    if logprobs_in_payload_condition:
        raise HTTPException(
            status_code=422,
            detail=(
                "Wrong type for logprobs."
                "If logprobs is set it should be an 32bit integer."
            ),
        )

    openai_top_logprobs_in_payload_condition = (
        "top_logprobs" in payload and not isinstace(payload["logprobs"], int)
    )

    if openai_top_logprobs_in_payload_condition: 
        raise HTTPException(
            status_code=422,
            detail=(
                "Wrong type for top_logprobs."
                "If top_logprobs is set it should a 32bit integer."
            ),
        )

    return rename_payload_keys(payload, key_mapping)
```

Here's how I think it should be. Let me know if you disagree or agree :).